packages/node/src/db/BetterSQLite3DBAdapter.ts

```
@@ -0,0 +1,329 @@
+            };
+            this.iterateListeners((cb) => cb.tablesUpdated?.(event));
+          }
+        }
+      } finally {
+        const next = this.writeQueue.shift();
+        if (next) {
+          next();
+        } else {
+          this.writeConnection.isBusy = false;
+        }
+      }
+    })();
+  }
+
+  readTransaction<T>(
+    fn: (tx: BetterSQLite3Transaction) => Promise<T>,
+    _options?: DBLockOptions | undefined
+  ): Promise<T> {
+    return this.readLock((ctx) => this.internalTransaction(ctx as RemoteConnection, fn));
+  }
+
+  writeTransaction<T>(
+    fn: (tx: BetterSQLite3Transaction) => Promise<T>,
+    _options?: DBLockOptions | undefined
+  ): Promise<T> {
+    return this.writeLock((ctx) => this.internalTransaction(ctx as RemoteConnection, fn));
+  }
+
+  private async internalTransaction<T>(
+    connection: RemoteConnection,
+    fn: (tx: BetterSQLite3Transaction) => Promise<T>
+  ): Promise<T> {
+    let finalized = false;
+    const commit = async (): Promise<QueryResult> => {
+      if (!finalized) {
+        finalized = true;
+        await connection.execute('COMMIT');
+      }
+      return { rowsAffected: 0 };
+    };
+    const rollback = async (): Promise<QueryResult> => {
+      if (!finalized) {
+        finalized = true;
+        await connection.execute('ROLLBACK');
+      }
+      return { rowsAffected: 0 };
+    };
+    try {
+      await connection.execute('BEGIN');
+      const result = await fn({
+        execute: (query, params) => connection.execute(query, params),
+        executeBatch: (query, params) => connection.executeBatch(query, params),
+        get: (query, params) => connection.get(query, params),
+        getAll: (query, params) => connection.getAll(query, params),
+        getOptional: (query, params) => connection.getOptional(query, params),
+        commit,
+        rollback
+      });
+      await commit();
+      return result;
+    } catch (ex) {
+      await rollback();
```

In some cases it can be necessary to swallow errors for this rollback. For [example](https://github.com/powersync-ja/powersync-js/blob/197a37a5fd1a6d74600a495234b575ed4e629d5b/packages/powersync-op-sqlite/src/db/OPSqliteAdapter.ts#L284).